src/radical/pilot/agent/launch_method/mpirun.py

```diff
@@ -15,114 +15,176 @@ class MPIRun(LaunchMethod):
-        if self.launch_command:
-            self.launch_command = os.path.basename(self.launch_command)
+        if '_dplace' in self.name.lower():
+            lm_info['dplace'] = ru.which('dplace')
+            assert lm_info['dplace']
 
+        # cheyenne always needs mpt and omplace
+        if 'cheyenne' in ru.get_hostname():
+            lm_info['omplace'] = ru.which('omplace')
+            lm_info['mpt']     = True
 
-        # do we need ccmrun or dplace?
-        if '_ccmrun' in self.name:
-            self._ccmrun = ru.which('ccmrun')
-            if not self._ccmrun:
-                raise RuntimeError("ccmrun not found!")
+        mpi_version, mpi_flavor = self._get_mpi_info(lm_info['command'])
+        lm_info['mpi_version']  = mpi_version
+        lm_info['mpi_flavor']   = mpi_flavor
 
-        if '_dplace' in self.name:
-            self._dplace = ru.which('dplace')
-            if not self._dplace:
-                raise RuntimeError("dplace not found!")
+        return lm_info
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _init_from_info(self, lm_info, lm_cfg):
+
+        self._env         = lm_info['env']
+        self._env_sh      = lm_info['env_sh']
+        self._command     = lm_info['command']
+
+        self._mpt         = lm_info['mpt']
+        self._rsh         = lm_info['rsh']
+        self._ccmrun      = lm_info['ccmrun']
+        self._dplace      = lm_info['dplace']
+        self._omplace     = lm_info['omplace']
 
-        self.mpi_version, self.mpi_flavor = \
-                                       self._get_mpi_info(self.launch_command)
+        self._mpi_version = lm_info['mpi_version']
+        self._mpi_flavor  = lm_info['mpi_flavor']
 
 
     # --------------------------------------------------------------------------
     #
-    def construct_command(self, t, launch_script_hop):
+    def finalize(self):
 
-        slots        = t['slots']
-        uid          = t['uid']
-        td          = t['description']
-        sandbox      = t['task_sandbox_path']
-        task_exec    = td['executable']
-        task_threads = td.get('cpu_threads', 1)
-        task_env     = td.get('environment') or dict()
-        task_args    = td.get('arguments')   or list()
-        task_argstr  = self._create_arg_string(task_args)
+        pass
 
-        if '_dplace' in self.name and task_threads > 1:
-            # dplace pinning would disallow threads to map to other cores
-            raise ValueError('dplace can not place threads [%d]' % task_threads)
 
-        # Construct the executable and arguments
-        if task_argstr: task_command = "%s %s" % (task_exec, task_argstr)
-        else          : task_command = task_exec
+    # --------------------------------------------------------------------------
+    #
+    def can_launch(self, task):
+
+        # mpirun can launch any executable
+        if task['description']['executable']:
```

Comment:
There is a different check in LM MPIExec, should for both of them be used a "combined" approach?
```python
        td = task['description']
        if td['executable'] and td['cpu_process_type'] == rpc.MPI:
            return True
```

Reply:
Thanks - I made the check the same, but for the time being would like to keep the two LMs separate.

Reply:
sorry, I didn't mean to merge them, but to have a check with both `executable` and `cpu_process_type` (probably no need to check for `executable`, since it always should be set)

Reply:
Well, the nice thing is that they also can launch non-MPI executables, so the test for `cpu_process_type` is actually not needed.  `executable` may not be set in the future for function based tasks.  So yeah, this is a bit premature, but ok I hope...